File: monitoring-service/infrastructure/tracing.py
"""
OpenTelemetry tracing setup for the HealthBridge monitoring service.

Single Responsibility — this module owns *only* tracing initialization.
The Flask app calls init_tracing(app) once at startup.

Auto-instruments:
  - Flask incoming requests (one span per HTTP request)
  - requests library outgoing calls (one span per downstream call)

Trace context (W3C tracecontext headers) flows automatically between services,
so a single user request shows up as a single trace spanning gateway → C# → Python.
"""

import os
import logging

from opentelemetry import trace
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.instrumentation.flask import FlaskInstrumentor
from opentelemetry.instrumentation.requests import RequestsInstrumentor

from core.constants import SERVICE_MONITORING

logger = logging.getLogger(__name__)

SERVICE_NAME = SERVICE_MONITORING
SERVICE_VERSION = "1.0.0"


def init_tracing(app):
    """
    Initialize OpenTelemetry tracing for the Flask app.

    The OTLP endpoint is configurable via OTEL_EXPORTER_OTLP_ENDPOINT
    (default: http://jaeger:4317 — matches the docker-compose service name).
    """
    endpoint = os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT", "http://jaeger:4317")

    # Resource describes this service to the trace backend
    resource = Resource.create({
        "service.name": SERVICE_NAME,
        "service.version": SERVICE_VERSION,
    })

    # Tracer provider — batches spans and exports via OTLP gRPC
    provider = TracerProvider(resource=resource)
    exporter = OTLPSpanExporter(endpoint=endpoint, insecure=True)
    provider.add_span_processor(BatchSpanProcessor(exporter))

    # Register globally so instrumentors find it
    trace.set_tracer_provider(provider)

    # Auto-instrument Flask + requests library
    FlaskInstrumentor().instrument_app(app)
    RequestsInstrumentor().instrument()

    logger.info("OpenTelemetry tracing initialized — exporting to %s", endpoint)

File: gateway/tracing.go
// OpenTelemetry tracing setup for the HealthBridge gateway.
//
// Initializes a tracer that exports spans to an OTLP collector
// (Jaeger, Tempo, or AWS X-Ray via the OTel Collector). The endpoint
// is configurable via OTEL_EXPORTER_OTLP_ENDPOINT, defaulting to the
// Jaeger service name in docker-compose.
//
// Single Responsibility — this file owns *only* tracing setup.
// main.go just calls initTracer() at startup.

package main

import (
	"context"
	"log/slog"
	"os"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.24.0"
)

// serviceName is what this process calls itself in traces. It is the same value
// the /health response reports, so a span in Jaeger and a health entry line up.
const serviceName = ServiceGateway

// initTracer configures the global OpenTelemetry tracer provider with an
// OTLP gRPC exporter. Returns a shutdown function that should be deferred
// from main() so spans are flushed before the process exits.
func initTracer() func(context.Context) error {
	endpoint := os.Getenv("OTEL_EXPORTER_OTLP_ENDPOINT")
	if endpoint == "" {
		endpoint = "jaeger:4317"
	}

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	// OTLP gRPC exporter — sends spans over the wire to a collector
	exporter, err := otlptrace.New(ctx,
		otlptracegrpc.NewClient(
			otlptracegrpc.WithEndpoint(endpoint),
			otlptracegrpc.WithInsecure(),
		),
	)
	if err != nil {
		slog.Warn("Failed to create OTLP exporter — continuing without tracing", "error", err)
		return func(context.Context) error { return nil }
	}

	// Resource describes this service to the trace backend
	res, err := resource.New(ctx,
		resource.WithAttributes(
			semconv.ServiceName(serviceName),
			semconv.ServiceVersion("1.0.0"),
		),
	)
	if err != nil {
		slog.Warn("Failed to create resource:", "error", err)
	}

	// Tracer provider — batches spans and forwards them to the exporter
	tp := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(exporter),
		sdktrace.WithResource(res),
		sdktrace.WithSampler(sdktrace.AlwaysSample()),
	)

	// Register globally so otelhttp middleware picks it up
	otel.SetTracerProvider(tp)

	slog.Info("OpenTelemetry tracing initialized", "endpoint", endpoint)

	return tp.Shutdown
}
